### DISCUSSION TITLE
Allow to reach out a service on any port

### DISCUSSION THREAD

**Discussion by AlesKrajnik** [istio/istio#49449] [Q&A] [answered]
Hi,

I am writing to ask the community for a help regarding my use-case.

I have two services that need to talk over TCP on multiple ports, however, the port numbers are not known upfront. Potentially I am talking about hundreds open ports. As far as I understand, in plain K8s it's doable because it's possible to communicate also undefined ports on services, but when istio is involved, all the ports must be explicitly defined.

The reason why the ports are not known is because they are being open dynamically based on configuration later after pods are created. Also, the ports change during runtime. So, there's a `serviceA` talking to `serviceB` on ports somewhere in the range 10000-19999. Another service, the `configurator` service, configures the `serviceB` to listen on various ports in the range and then configures the `serviceA` to use those ports.

A solution would be to allow a port range on a service to be used by istio. Do I get it right that the issue https://github.com/istio/istio/issues/23983 aims at exactly that, meaning that a port range is currently not possible in istio?

Other solutions I explored:

- all communication is multiplexed via a single TCP port - AFAIK the software running in the containers does not allow that, mostly because each port is bound to different UID
- each port should be services by it's own POD - such a solution wouldn't scale well for my use-case, as there might be hundreds of open ports
- communicate using UNIX socket over a shared volume - that works, that's what do I now, but it requires a shared volume which in turn requires all the pods require on the same node; TCP would be preferred

Any idea would be appreciated, thank you!

**Comment by howardjohn** [comment]
> in plain K8s it's doable because it's possible to communicate also undefined ports on services,

This is not the case AFAIK; if it was then Istio would definitely prioritize this 'any port' use case

**Comment by AlesKrajnik** [comment]
@howardjohn Sorry, you're right. Not sure, where I got that from, but it doesn't make sense indeed.

So if I get it right that it means there's no way to define a port range, not in K8s in general (https://github.com/kubernetes/kubernetes/issues/23864) and therefore also not in istio?

Thanks!

**Reply by howardjohn** [reply, marked as answer]
Correct. In theory, Istio could do it without K8s native support (via ServiceEntry, probably), but it hasn't. Largely this is due to it being very hard to implement in Istio's current architecture

**Reply by AlesKrajnik** [reply]
Thank you!
